File: dotnet/src/Core/TopicGenerator.cs
namespace Agience.Core
{
    public class TopicGenerator
    {

        public const string EVENT_PREFIX =  "event/";
        public const string CONNECT_PREFIX = "connect/";

        private readonly string _authorityId;
        private readonly string _senderId;

        public TopicGenerator(string authorityId, string senderId)
        {
            _authorityId = authorityId ?? throw new ArgumentNullException(nameof(authorityId));
            _senderId = senderId == authorityId ? "-" : senderId ?? throw new ArgumentNullException(nameof(senderId));
        }

        public string ConnectTo(string topic)
        {
            return $"{CONNECT_PREFIX}{topic}";
        }

        public string SubscribeAs(string? hostId, string? agentId)
        {
            return $"{EVENT_PREFIX}+/{_authorityId}/{hostId ?? "-"}/{agentId ?? "-"}";
        }

        public string PublishTo(string? hostId, string? agentId)
        {
            return $"{EVENT_PREFIX}{_senderId}/{_authorityId}/{hostId ?? "-"}/{agentId ?? "-"}";
        }


        public string SubscribeAsAgent()
        {
            return SubscribeAs(null, _senderId);
        }

        public string PublishToAgent(string? agentId)
        {
            return PublishTo(null, agentId);
        }

        public string PublishToHost(string hostId)
        {
            return PublishTo(hostId, null);
        }

        public string SubscribeAsHost()
        {
            return SubscribeAs(_senderId, null);
        }

        public string PublishToAuthority()
        {
            return PublishTo(null, null);
        }

        public string SubscribeAsAuthority()
        {
            return SubscribeAs(null, null);
        }
    }
}

File: python/src/core/topic_generator.py
class TopicGenerator:
    EVENT_PREFIX = "event/"
    CONNECT_PREFIX = "connect/"

    def __init__(self, authority_id: str, sender_id: str):
        if authority_id is None:
            raise ValueError("authority_id cannot be None")
        if sender_id is None:
            raise ValueError("sender_id cannot be None")

        self._authority_id = authority_id
        self._sender_id = "-" if sender_id == authority_id else sender_id

    def connect_to(self, topic: str) -> str:
        return f"{self.CONNECT_PREFIX}{topic}"

    def subscribe_as(self, host_id: str | None, agent_id: str | None) -> str:
        return f"{self.EVENT_PREFIX}+/{self._authority_id}/{host_id or '-'}/{agent_id or '-'}"

    def publish_to(self, host_id: str | None, agent_id: str | None) -> str:
        return f"{self.EVENT_PREFIX}{self._sender_id}/{self._authority_id}/{host_id or '-'}/{agent_id or '-'}"

    def subscribe_as_agent(self) -> str:
        return self.subscribe_as(None, self._sender_id)

    def publish_to_agent(self, agent_id: str | None) -> str:
        return self.publish_to(None, agent_id)

    def publish_to_host(self, host_id: str) -> str:
        return self.publish_to(host_id, None)

    def subscribe_as_host(self) -> str:
        return self.subscribe_as(self._sender_id, None)

    def publish_to_authority(self) -> str:
        return self.publish_to(None, None)

    def subscribe_as_authority(self) -> str:
        return self.subscribe_as(None, None)
